### DISCUSSION TITLE
Gas sensor + GPIO: Error compelling for board ESP32 Wrover Module

### DISCUSSION THREAD

**Discussion by bramboe** [HomeSpan/HomeSpan#74] [Q&A]
I'm trying to understand the HomeSpan construction of building my gas sensor script based on the examples I found.
I get the following error: Error compelling for board ESP32 Wrover Module (see below), I can't seem to figure out what the issue is. Can someone point me in the right direction?

What I'm trying to do:

- Add the sensor
- Set characteristics within the Struct
- Setting the pin Int 
- Make sure the digitalWrite function to change the value
- Setting the accessory information
- Set the GPIO 
- And in the loop I want the sensor to be updated.

Script (ignore the comments, I have not changed them from the example files)

    #include "HomeSpan.h"         // Always start by including the HomeSpan library
  
    struct CarbonSensor : Service::CarbonDioxideSensor {     // A standalone Air Quality sensor
  
  
    // An Air Quality Sensor is similar to a Temperature Sensor except that it supports a wide variety of measurements.
    // We will use three of them.  The first is required, the second two are optional.
    int carbonPin;                               // store the pin number connected to a hypothetical relay that turns the Table Lamp on/off
  
    SpanCharacteristic *statActive;                 // reference to the Air Quality Characteristic, which is an integer from 0 to 5
    SpanCharacteristic *carbonDioLvl;                  // reference to the Ozone Density Characteristic, which is a float from 0 to 1000
    SpanCharacteristic *carbonDioPeak;                 // reference to the Nitrogen Dioxide Characteristic, which is a float from 0 to 1000
    
    CarbonSensor(int carbonPin) : Service::CarbonDioxideSensor(){       // constructor() method
  
      statActive=new Characteristic::StatusActive(1);                         // instantiate the Air Quality Characteristic and set initial value to 1
      carbonDioLvl=new Characteristic::CarbonDioxideLevel(300.0);                    // instantiate the Ozone Density Characteristic and set initial value to 300.0
      carbonDioPeak=new Characteristic::CarbonDioxidePeakLevel(700.0);         // instantiate the Nitrogen Dioxide Density Characteristic and set initial value to 700.0
  
      this->carbonPin=carbonPin;                   // save the pin number for the hypothetical relay
      pinMode(carbonPin,OUTPUT);                 // configure the pin as an output using the standard Arduino pinMode function      
  
    }  
    boolean update(){                          // update() method
  
      digitalWrite(carbonPin,statActive->getNewVal());      // use standard Arduino digitalWrite function to change the ledPin to either high or low based on the value requested by HomeKit
      digitalWrite(carbonPin,carbonDioLvl->getNewVal());      // use standard Arduino digitalWrite function to change the ledPin to either high or low based on the value requested by HomeKit
      digitalWrite(carbonPin,carbonDioPeak->getNewVal());      // use standard Arduino digitalWrite function to change the ledPin to either high or low based on the value requested by HomeKit
  
  
     
      return(true);                            // return true to let HomeKit (and the Home App Client) know the update was successful
    
      } // end update()
  
    void setup() {
    
      // Example 3 shows how adding multiple Services to a single Accessory allows us to create a multi-featured Accessory, such as a ceiling fan wih a ceiling light
     
      Serial.begin(115200);      // Start a serial connection - this is needed for you to type in your WiFi credentials
    
      homeSpan.begin(Category::Sensors,"Sensors");  // Begin a HomeSpan Session - note the Category has been set to "Fans"
    
      // We begin by creating a Light Bulb Accessory just as in Examples 1 and 2, but with Name now set to "My Ceiling Fan"
    
      new SpanAccessory();                            // Begin by creating a new Accessory using SpanAccessory(), which takes no arguments 
      
        new Service::AccessoryInformation();            // HAP requires every Accessory to implement an AccessoryInformation Service, which has 6 required Characteristics
          new Characteristic::Name("Sensor");     // Name of the Accessory, which shows up on the HomeKit "tiles", and should be unique across Accessories
          new Characteristic::Manufacturer("Devices");   // Manufacturer of the Accessory (arbitrary text string, and can be the same for every Accessory)
          new Characteristic::SerialNumber("123-ABC");    // Serial Number of the Accessory (arbitrary text string, and can be the same for every Accessory)
          new Characteristic::Model("5-Volt Sensor");     // Model of the Accessory (arbitrary text string, and can be the same for every Accessory)
          new Characteristic::FirmwareRevision("0.9");    // Firmware of the Accessory (arbitrary text string, and can be the same for every Accessory) 
          new Characteristic::Identify();                 // Create the required Identify  
          
        new Service::HAPProtocolInformation();          // Create the HAP Protcol Information Service  
          new Characteristic::Version("1.1.0");           // Set the Version Characteristicto "1.1.0" as required by HAP
    
       new CarbonSensor(4);                               // instantiate the TableLamp Service (defined below) with lampPin set to 17
    
        new Service::CarbonDioxideSensor();                       // Create the Light Bulb Service
          new Characteristic::CarbonDioxideDetected();                       // This Service requires the "On" Characterstic to turn the light on and off
    
      // IMPORTANT:  HomeKit Controllers often cache a lot of information.  If your Controller does not update to match the above configuration,
      // simply select the Accessory in your Controller and under setting, select "Remove Accessory" and then re-pair.
    
    } // end of setup()
    
    
    
    void loop(){
    
    //       Note we are NOT updating the Nitrogen Dioxide Density Characteristic.  This should therefore remain steady at its initial value of 700.0
    
        if(carbonDioLvl->timeVal()>5000)                            // modify the Air Quality Characteristic every 5 seconds
          carbonDioLvl->setVal((carbonDioLvl->getVal()+1)%6);         // simulate a change in Air Quality by incrementing the current value by one, and keeping in range 0-5
    
        if(carbonDioPeak->timeVal()>10000)                            // modify the Ozone Density Characteristic value every 10 seconds
          carbonDioPeak->setVal((double)random(200,500));             // simulate a change with a random value between 200 and 499.  Note use of (double) cast since random() returns an integer
    
     homeSpan.poll();         // run HomeSpan!
     } 
    };  `

Error
    
    Arduino: 1.8.13 (Mac OS X), Board: "ESP32 Wrover Module, Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS), QIO, 80MHz, 115200, None"

            /var/folders/2q/h3zhblhj6611x4r9x0_b4_2w0000gp/T/arduino_cache_668355/core/core_4f015046e1a2423cc3269e91996c6b66.a(m    ain.cpp.o):(.literal._Z8loopTaskPv+0x8): undefined reference to 'setup()'
    /var/folders/2q/h3zhblhj6611x4r9x0_b4_2w0000gp/T/arduino_cache_668355/core/core_4f015046e1a2423cc3269e91996c6b66.a(m    ain.cpp.o):(.literal._Z8loopTaskPv+0xc): undefined reference to 'loop()'
    /var/folders/2q/h3zhblhj6611x4r9x0_b4_2w0000gp/T/arduino_cache_668355/core/core_4f015046e1a2423cc3269e91996c6b66.a(main.cpp.o): In function 'loopTask(void*)':
    /Users/[name]/Documents/Arduino/hardware/espressif/esp32/cores/esp32/main.cpp:18: undefined reference to 'setup()'
    Multiple libraries were found for "WiFi.h"
    Used: /Users/[name]/Documents/Arduino/hardware/espressif/esp32/libraries/WiFi
    Not used: /private/var/folders/2q/h3zhblhj6611x4r9x0_b4_2w0000gp/T/AppTranslocation/4B695B46-37D9-4FE0-887D-11B8F7B67A58/d/Arduino.app/Contents/Java/libraries/WiFi
    /Users/[name]/Documents/Arduino/hardware/espressif/esp32/cores/esp32/main.cpp:21: undefined reference to 'loop()'
    collect2: error: ld returned 1 exit status
    exit status 1
    Error compiling for board ESP32 Wrover Module.
    Error while setting serial port parameters: 115,200 N 8 1
    Error while setting serial port parameters: 115,200 N 8 1
    Error while setting serial port parameters: 115,200 N 8 1
    Error while setting serial port parameters: 115,200 N 8 1
    Error while setting serial port parameters: 115,200 N 8 1
`

**Comment by HomeSpan** [comment]
It looks like you've tried to encapsulate all your Arduino Code into the Carbon Sensor structure, and you've also combined the Arduino loop() _function_ with the Carbon Sensor loop() _method_, as well as embedded the Arduino setup() function in the sensor as a method of the structure.  Every Arduino sketch needs standalone setup() and loop() _functions_.  Your Carbon Sensor should not contain these global functions, else they become methods of the structure (the compiler error above is due to the fact that the compiler cannot find the Arduino's main loop() function).  Note the Carbon Sensor can have its own loop() method, but this is distinct from the main Arduino loop() function.

In addition to reviewing the examples, you may want to review the [API Overview](https://github.com/HomeSpan/HomeSpan/blob/master/docs/Overview.md) to see how a typical HomeSpan sketch is structured.
